### DISCUSSION TITLE
esp_dmx v4.0.0

### DISCUSSION THREAD

**Discussion by someweisguy** [someweisguy/esp_dmx#125] [Announcements]
This release adds some more functionality to the RDM responder including RDM queues and RDM sensors as well as updated internal logic to better handle some edge cases of RDM. Because these edge cases (such as RDM queues and sub-devices) are so unusual, the ABI has been retooled to make the user experience easier. The goal is to remain on v4 for a while so that few, if any, breaking changes will need to be made for future updates. 

Some future goals of v4 include:

- Unit tests to make sure that the code base is robust
- Parity between RDM responder and controller functions; right now there are far more RDM responder functions than controller functions
- Stronger documentation for advanced features, such as the custom RDM parameter ABI
- Support for sending and receiving RDM ACK overflow and timer messages
- Base support for more RDM parameters, both requests and responses
- Support for multiple sub-devices

Please reach out if there are any questions, comments, or concerns!

## What's Changed
* Fix GPTimer error message spam by @someweisguy in https://github.com/someweisguy/esp_dmx/pull/68
* Fix ESP32-C6 compatiblity by @someweisguy in https://github.com/someweisguy/esp_dmx/pull/97
* add missing ISR_ATTR to methods called from isrs by @arneboe in https://github.com/someweisguy/esp_dmx/pull/98
* Implemented personality and personality_description parameters by @arneboe in https://github.com/someweisguy/esp_dmx/pull/92
* Add device label parameter by @arneboe in https://github.com/someweisguy/esp_dmx/pull/93
* Add PARAMETER_DESCRIPTION and a way to define custom parameters by @arneboe in https://github.com/someweisguy/esp_dmx/pull/100
* Update HAL files by @someweisguy in https://github.com/someweisguy/esp_dmx/pull/106
* Add RDM queue by @someweisguy in https://github.com/someweisguy/esp_dmx/pull/102
* Update File Structure by @someweisguy in https://github.com/someweisguy/esp_dmx/pull/107
* Fix UART Logic by @someweisguy in https://github.com/someweisguy/esp_dmx/pull/122
* Feature/rdm-response-function by @someweisguy in https://github.com/someweisguy/esp_dmx/pull/123

**Full Changelog**: https://github.com/someweisguy/esp_dmx/compare/v3.1.0...v4.0.0

<hr /><em>This discussion was created from the release <a href='https://github.com/someweisguy/esp_dmx/releases/tag/v4.0.0'>esp_dmx v4.0.0</a>.</em>

**Comment by HopingCoding** [comment]
Congratulations on the release!
I tried to upgrade but I am getting a compile error:

```
.pio/libdeps/upesy_wroom/esp_dmx/src/dmx/hal/timer.c: In function 'dmx_timer_init':
.pio/libdeps/upesy_wroom/esp_dmx/src/dmx/hal/timer.c:97:52: error: 'isr_handle' undeclared (first use in this function); did you mean 'intr_handle_t'?
   timer_isr_callback_add(timer->group, timer->idx, isr_handle, isr_context,
                                                    ^~~~~~~~~~
                                                    intr_handle_t
```

My board is a ESP32 WROOM and I am using the Arduino environment if that is relevant.

**Reply by someweisguy** [reply]
Thanks for bringing this to my attention! I am not sure how this got through my tests; I thought I was able to compile on Arduino when I last tested. But when I just tried to compile again this evening, I was able to reproduce the issue. 

This is fixed in #126. I'll have a patch out shortly - feel free to clone the update into your Arduino libraries folder. Or you can wait until Arduino pushes the update to the Arduino Library Manager. :)
